File: tests/test_presets_when.py
"""Value Presets: `when` conditions and `#` hints.

Runs without ComfyUI: `nodes` is stubbed the way test_security.py stubs
it, so borrowed dropdowns can be declared per test.
"""

import importlib
import sys
import types
import unittest
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
if "obvpm_testpack" not in sys.modules:
    pack = types.ModuleType("obvpm_testpack")
    pack.__path__ = [str(ROOT)]
    sys.modules[pack.__name__] = pack
try:
    import torch  # noqa: F401
except ImportError:
    sys.modules.setdefault("torch", types.ModuleType("torch"))
if "nodes" not in sys.modules:
    stub = types.ModuleType("nodes")
    stub.NODE_CLASS_MAPPINGS = {}
    sys.modules["nodes"] = stub

presets = importlib.import_module("obvpm_testpack.presets")


def install(**classes):
    """Make these the installed node classes.

    Looked up at call time: test_security.py REPLACES the `nodes` stub
    at import, and under discover it is imported after this file.
    """
    mapping = sys.modules["nodes"].NODE_CLASS_MAPPINGS
    mapping.clear()
    mapping.update(classes)

SCHEMA = """
turbo_loader: choice off, normal, larryvrh = off   # which loader applies the LoRA
turbo_lora: @Loader.lora_name when turbo_loader != off  # the LoRA file
turbo_strength: float 0..1.00 = 1.0 when turbo_loader != off
spectrum: bool = false
spectrum_amount: int 0..10 = 3 when spectrum = true
steps: int = 20
"""


class Loader:
    @classmethod
    def INPUT_TYPES(cls):
        return {"required": {"lora_name": (["a.safetensors", "b.safetensors"],)}}


class WhenParsing(unittest.TestCase):
    def setUp(self):
        install(Loader=Loader)

    def test_tails_come_off_in_order(self):
        fields = {f.name: f for f in presets.parse_schema(SCHEMA)}
        self.assertEqual(fields["turbo_loader"].hint, "which loader applies the LoRA")
        self.assertIsNone(fields["turbo_loader"].when)
        self.assertEqual(fields["turbo_loader"].default_text, "off")
        lora = fields["turbo_lora"]
        self.assertEqual(lora.ref, "Loader.lora_name")
        self.assertEqual(lora.hint, "the LoRA file")
        self.assertEqual((lora.when.field, lora.when.negate, lora.when.values),
                         ("turbo_loader", True, ["off"]))
        self.assertEqual(lora.when.text, "turbo_loader != off")
        strength = fields["turbo_strength"]
        self.assertEqual((strength.lo, strength.hi, strength.default_text, strength.span_text),
                         (0.0, 1.0, "1.0", "0..1.00"))
        self.assertEqual(strength.when.text, "turbo_loader != off")
        self.assertEqual(fields["spectrum_amount"].when.values, ["true"])
        self.assertEqual(fields["steps"].hint, "")

    def test_hint_may_say_when_and_a_default_may_be_anything_else(self):
        f = presets.parse_schema("a: text = go #b # when it is time")[0]
        self.assertEqual((f.default_text, f.hint), ("go", "b # when it is time"))
        f = presets.parse_schema("mode: choice fast, when, slow = when")[0]
        self.assertEqual((f.choices(), f.default_text), (["fast", "when", "slow"], "when"))
        f = presets.parse_schema("a#b: int = 3 #the hint")[0]
        self.assertEqual((f.name, f.hint), ("a#b", "the hint"))

    def test_bool_values_normalise_and_any_of_lists(self):
        f = presets.parse_schema("s: bool\nx: int when s = on, no")[1]
        self.assertEqual(f.when.values, ["true", "false"])
        self.assertEqual(f.when.text, "s = true, false")
        f = presets.parse_schema("m: choice a, b, c\nx: int when m = a, c")[1]
        self.assertTrue(f.when.holds("a"))
        self.assertTrue(f.when.holds("c"))
        self.assertFalse(f.when.holds("b"))
        f = presets.parse_schema("m: choice a, b, c\nx: int when m != a, c")[1]
        self.assertFalse(f.when.holds("a"))
        self.assertTrue(f.when.holds("b"))

    def test_refusals_name_the_line(self):
        cases = [
            ("x: int when y = 1", "not a field declared above"),
            ("x: int when x = 1", "depends on itself"),
            ("y: int\nx: int when y = 1", "only a choice or bool"),
            ("y: text\nx: int when y = a", "only a choice or bool"),
            ("y: choice a, b\nx: int when y = c", "not one of its choices"),
            ("y: choice a, b\nx: int when y", "not written"),
            ("y: choice a, b\nx: int when y =", "not written"),
            ("y: bool\nx: int when y = maybe", "not true or false"),
            ("x: int when y = 1\ny: choice 1, 2", "not a field declared above"),
        ]
        for schema, why in cases:
            with self.assertRaises(presets.SchemaError, msg=schema) as caught:
                presets.parse_schema(schema)
            self.assertIn(why, str(caught.exception), schema)
            self.assertIn("schema line", str(caught.exception))

    def test_a_borrowed_list_is_checked_at_run_time_not_parse_time(self):
        # the loader's list is live, so the parser cannot know it
        presets.parse_schema("l: @Loader.lora_name\nx: int when l = z.safetensors")


class WhenResolving(unittest.TestCase):
    def setUp(self):
        install(Loader=Loader)

    def test_hidden_fields_are_none_whatever_is_stored(self):
        packed, _ = presets.resolve(
            SCHEMA, '{"turbo_loader": "off", "turbo_lora": "a.safetensors", '
                    '"turbo_strength": 0.5, "spectrum": false, "spectrum_amount": 7}')
        self.assertEqual(packed, {
            "turbo_loader": "off", "turbo_lora": None, "turbo_strength": None,
            "spectrum": False, "spectrum_amount": None, "steps": 20})
        # still in schema order, hidden or not: names on the bundle are
        # what Unbundle traces
        self.assertEqual(list(packed), ["turbo_loader", "turbo_lora", "turbo_strength",
                                        "spectrum", "spectrum_amount", "steps"])

    def test_shown_fields_carry_their_values(self):
        packed, _ = presets.resolve(
            SCHEMA, '{"turbo_loader": "normal", "turbo_lora": "b.safetensors", '
                    '"turbo_strength": 0.5, "spectrum": true, "spectrum_amount": 7}')
        self.assertEqual(packed["turbo_lora"], "b.safetensors")
        self.assertEqual(packed["turbo_strength"], 0.5)
        self.assertEqual(packed["spectrum_amount"], 7)

    def test_a_hidden_field_is_not_coerced(self):
        # a deleted LoRA behind an off switch must not refuse the run...
        packed, _ = presets.resolve(
            SCHEMA, '{"turbo_loader": "off", "turbo_lora": "gone.safetensors"}')
        self.assertIsNone(packed["turbo_lora"])
        # ... but the same value shown is still refused
        with self.assertRaises(ValueError):
            presets.resolve(SCHEMA, '{"turbo_loader": "normal", "turbo_lora": "gone.safetensors"}')

    def test_the_default_decides_when_nothing_is_stored(self):
        packed, _ = presets.resolve("s: bool = true\nx: int = 4 when s = true", "{}")
        self.assertEqual(packed, {"s": True, "x": 4})
        packed, _ = presets.resolve("s: bool\nx: int = 4 when s = true", "{}")
        self.assertEqual(packed, {"s": False, "x": None})

    def test_a_field_decided_by_a_hidden_field_is_hidden(self):
        schema = ("a: bool = false\n"
                  "b: choice x, y = y when a = true\n"
                  "c: int = 1 when b != x\n")
        # b is hidden (a is false), so c is hidden although b != x holds
        packed, _ = presets.resolve(schema, "{}")
        self.assertEqual(packed, {"a": False, "b": None, "c": None})
        packed, _ = presets.resolve(schema, '{"a": true}')
        self.assertEqual(packed, {"a": True, "b": "y", "c": 1})

    def test_a_numeric_choice_decides_by_its_text(self):
        install(Depth=type("Depth", (), {"INPUT_TYPES": classmethod(
            lambda cls: {"required": {"bit_depth": (["auto", 8, 10],)}})}))
        schema = "depth: @Depth.bit_depth\nfast: bool = true when depth = 8, 10"
        packed, _ = presets.resolve(schema, '{"depth": "8"}')
        self.assertEqual(packed, {"depth": 8, "fast": True})
        packed, _ = presets.resolve(schema, '{"depth": "auto"}')
        self.assertEqual(packed, {"depth": "auto", "fast": None})

    def test_describe_carries_the_condition_and_the_hint(self):
        described = {f["name"]: f for f in presets.describe(SCHEMA)}
        lora = described["turbo_lora"]
        self.assertEqual(lora["when"], {"field": "turbo_loader", "not": True, "values": ["off"]})
        self.assertEqual(lora["when_text"], "turbo_loader != off")
        self.assertEqual(lora["hint"], "the LoRA file")
        self.assertEqual(lora["choices"], ["a.safetensors", "b.safetensors"])
        self.assertIsNone(described["steps"]["when"])
        self.assertEqual(described["steps"]["when_text"], "")
        self.assertEqual(described["spectrum_amount"]["when"]["values"], ["true"])

    def test_the_label_check_ignores_hidden_fields(self):
        library = '{"p": {"turbo_loader": "off", "turbo_lora": "a.safetensors", "steps": 20}}'
        with self.assertLogs("obvpm", level="INFO") as logged:
            presets._LOG.info("marker")
            presets.resolve(SCHEMA, '{"turbo_loader": "off", "turbo_lora": "b.safetensors"}',
                            library, "p")
        self.assertFalse([m for m in logged.output if "changed" in m], logged.output)
        with self.assertLogs("obvpm", level="INFO") as logged:
            presets.resolve(SCHEMA, '{"turbo_loader": "off", "steps": 8}', library, "p")
        self.assertTrue([m for m in logged.output if "steps changed" in m], logged.output)


if __name__ == "__main__":
    unittest.main()

File: tests/test_presets_when.mjs
// Value Presets: conditional fields on the node's face, hints, and the
// schema editor's line round-trip. Runs the real web/value_presets.js in
// a vm with the framework mocked and the server's `describe` answered
// from a fixed field list -- the Python side (test_presets_when.py) owns
// the parsing; this checks what the browser does with what it is told.
//
// Run: node --experimental-vm-modules --test tests/test_presets_when.mjs
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import vm from "node:vm";
import test from "node:test";

const root = new URL("../web/", import.meta.url);
const tick = () => new Promise((resolve) => setImmediate(resolve));

// What /obvpm/presets/schema would answer for the turbo schema.
const FIELDS = [
    { name: "turbo_loader", kind: "choice", ref: null, lo: null, hi: null, span: "",
      decimals: null, default_text: "off", choices: ["off", "normal", "larryvrh"],
      default: "off", hint: "which loader applies the LoRA", when: null, when_text: "" },
    { name: "turbo_lora", kind: "choice", ref: "LoraName (obvpm).lora_name", lo: null,
      hi: null, span: "", decimals: null, default_text: "", choices: ["a.safetensors", "b.safetensors"],
      default: "a.safetensors", hint: "the LoRA file",
      when: { field: "turbo_loader", not: true, values: ["off"] }, when_text: "turbo_loader != off" },
    { name: "turbo_strength", kind: "float", ref: null, lo: 0, hi: 1, span: "0..1.00",
      decimals: 2, default_text: "1.0", choices: null, default: 1, hint: "",
      when: { field: "turbo_loader", not: true, values: ["off"] }, when_text: "turbo_loader != off" },
    { name: "spectrum", kind: "bool", ref: null, lo: null, hi: null, span: "", decimals: null,
      default_text: "false", choices: null, default: false, hint: "", when: null, when_text: "" },
    { name: "spectrum_amount", kind: "int", ref: null, lo: 0, hi: 10, span: "0..10", decimals: 0,
      default_text: "3", choices: null, default: 3, hint: "",
      when: { field: "spectrum", not: false, values: ["true"] }, when_text: "spectrum = true" },
    { name: "steps", kind: "int", ref: null, lo: null, hi: null, span: "", decimals: 0,
      default_text: "20", choices: null, default: 20, hint: "", when: null, when_text: "" },
];

async function harness() {
    const extensions = [];
    const app = { registerExtension: (ext) => extensions.push(ext), canvas: {} };
    const asked = { schema: 0 };
    const api = { fetchApi: async (url) => {
        if (url === "/obvpm/presets/schema") {
            asked.schema += 1;
            return { json: async () => ({ fields: FIELDS }) };
        }
        throw new Error("Offline test: " + url);
    } };
    const context = vm.createContext({
        console, setTimeout, clearTimeout, setImmediate,
        document: { body: {}, createElement: () => ({ style: {}, addEventListener() {},
            appendChild() {}, append() {} }), documentElement: {}, addEventListener() {} },
        window: {}, LiteGraph: {}, MutationObserver: class { observe() {} disconnect() {} },
        navigator: {},
    });
    const noop = () => {};
    const cache = new Map();
    const mock = (values) => new vm.SyntheticModule(Object.keys(values), function () {
        for (const [name, value] of Object.entries(values)) this.setExport(name, value);
    }, { context });
    async function load(name) {
        if (cache.has(name)) return cache.get(name);
        let mod;
        if (name.endsWith("scripts/app.js")) mod = mock({ app });
        else if (name.endsWith("scripts/api.js")) mod = mock({ api });
        else if (name === "obvpm_ui.js") {
            mod = mock({ themePalette: () => ({}), el: noop,
                valueTooltip: (w, width, base) => base,
                ...Object.fromEntries(["TEXT", "TITLE", "INK", "DIM", "EDGE", "FILL", "PANEL",
                    "textBox", "pushButton", "openOverlay", "dropWidgetSockets",
                    "askText", "askConfirm", "notice", "NODE_BUTTON", "NODE_BUTTON_HOVER",
                    "nodeButton", "nodeButtonBar", "paintNodeButton"].map((k) => [k, noop])) });
        } else {
            let source = await readFile(new URL(name, root), "utf8");
            if (name === "value_presets.js") {
                source += "\nexport { rebuild, isShown, applyVisibility, rowOf, lineOf, schemaOf, readJson, asDropdown };";
            }
            mod = new vm.SourceTextModule(source, { context, identifier: name });
        }
        cache.set(name, mod);
        await mod.link((spec) => load(spec.startsWith("./") ? spec.slice(2) : spec));
        return mod;
    }
    const mod = await load("value_presets.js");
    await mod.evaluate();
    return { p: mod.namespace, extensions, asked };
}

/** A litegraph-shaped node: enough for rebuild() to run against. */
class Node {
    constructor(values) {
        this.id = 1;
        this.size = [300, 100];
        this.widgets = [
            { name: "schema", value: "turbo_loader: choice off, normal, larryvrh = off" },
            { name: "preset", value: "custom", type: "combo", options: {} },
            { name: "values", value: JSON.stringify(values) },
            { name: "presets", value: "{}" },
            { name: "names", value: "" },
        ];
        this.inputs = [];
        this.graph = { getNodeById: () => this };
    }
    addWidget(type, name, value, callback, options) {
        const w = { type, name, value, callback, options: options ?? {} };
        this.widgets.push(w);
        // frontend 1.53: a widget added under a name the node already
        // has is renamed "name#1" on registration (ensureUniqueWidgetNames)
        if (this.widgets.some((o) => o !== w && o.name === name)) w.name = name + "#1";
        return w;
    }
    addDOMWidget(name, type, element, options) {
        const w = { name, type, element, options: options ?? {} };
        this.widgets.push(w);
        return w;
    }
    computeSize() {
        const shown = this.widgets.filter((w) => !w.hidden).length;
        return [this.size[0], 30 + shown * 24];
    }
    setSize(size) { this.size = size; }
    setDirtyCanvas() {}
}

function field(node, name) {
    return node.widgets.find((w) => w.name === name);
}

test("a false condition hides the widget and keeps its value", async () => {
    const { p } = await harness();
    const node = new Node({ turbo_loader: "off", turbo_lora: "b.safetensors",
                            turbo_strength: 0.5, spectrum: false, spectrum_amount: 7, steps: 20 });
    await p.rebuild(node, true);
    await tick();
    assert.equal(field(node, "turbo_lora").hidden, true);
    assert.equal(field(node, "turbo_lora").options.hidden, true);
    assert.equal(field(node, "turbo_strength").hidden, true);
    assert.equal(field(node, "spectrum_amount").hidden, true);
    assert.equal(!!field(node, "turbo_loader").hidden, false);
    assert.equal(!!field(node, "steps").hidden, false);
    // the value is still there, for when the condition holds again
    assert.equal(field(node, "turbo_lora").value, "b.safetensors");
    assert.equal(p.readJson(node, "values").turbo_lora, "b.safetensors");
    // and the node shrank to the widgets it shows
    assert.equal(node.size[1], node.computeSize()[1]);
});

test("changing the deciding widget shows and hides its dependents at once", async () => {
    const { p } = await harness();
    const node = new Node({ turbo_loader: "off", spectrum: false });
    await p.rebuild(node, true);
    await tick();
    assert.equal(field(node, "turbo_lora").hidden, true);
    const before = node.size[1];
    field(node, "turbo_loader").callback("normal");
    assert.equal(field(node, "turbo_lora").hidden, false);
    assert.equal(field(node, "turbo_lora").options.hidden, false);
    assert.equal(field(node, "turbo_strength").hidden, false);
    assert.ok(node.size[1] > before, "the node grew for the shown fields");
    assert.equal(field(node, "spectrum_amount").hidden, true);
    field(node, "spectrum").callback(true);
    assert.equal(field(node, "spectrum_amount").hidden, false);
    field(node, "turbo_loader").callback("off");
    assert.equal(field(node, "turbo_lora").hidden, true);
    assert.equal(field(node, "turbo_strength").hidden, true);
    assert.equal(field(node, "spectrum_amount").hidden, false);
});

test("a field decided by a hidden field is hidden", async () => {
    const { p } = await harness();
    const map = {
        a: { name: "a", kind: "bool", when: null },
        b: { name: "b", kind: "choice", when: { field: "a", not: false, values: ["true"] } },
        c: { name: "c", kind: "int", when: { field: "b", not: true, values: ["x"] } },
    };
    const shown = {};
    for (const f of Object.values(map)) shown[f.name] = p.isShown(f, map, { a: false, b: "y" }, shown);
    assert.deepEqual(shown, { a: true, b: false, c: false });
    const again = {};
    for (const f of Object.values(map)) again[f.name] = p.isShown(f, map, { a: true, b: "y" }, again);
    assert.deepEqual(again, { a: true, b: true, c: true });
    // a numeric choice decides by its text, as on the server
    const depth = { depth: { name: "depth", kind: "choice", when: null },
                    fast: { name: "fast", kind: "bool", when: { field: "depth", not: false, values: ["8", "10"] } } };
    assert.equal(p.isShown(depth.fast, depth, { depth: 8 }, { depth: true }), true);
    assert.equal(p.isShown(depth.fast, depth, { depth: "auto" }, { depth: true }), false);
});

test("the hint is the widget's tooltip, after an error when there is one", async () => {
    const { p } = await harness();
    const node = new Node({});
    await p.rebuild(node, true);
    await tick();
    assert.equal(field(node, "turbo_loader").tooltip, "which loader applies the LoRA");
    assert.equal(field(node, "steps").tooltip, undefined);
});

test("editor rows carry the condition and hint and round-trip to the line", async () => {
    const { p } = await harness();
    const rows = FIELDS.map(p.rowOf);
    assert.equal(rows[1].when, "turbo_loader != off");
    assert.equal(rows[1].hint, "the LoRA file");
    assert.equal(p.lineOf(rows[0]),
                 "turbo_loader: choice off, normal, larryvrh = off # which loader applies the LoRA");
    assert.equal(p.lineOf(rows[1]),
                 "turbo_lora: @LoraName (obvpm).lora_name when turbo_loader != off # the LoRA file");
    assert.equal(p.lineOf(rows[2]), "turbo_strength: float 0..1.00 = 1.0 when turbo_loader != off");
    assert.equal(p.lineOf(rows[4]), "spectrum_amount: int 0..10 = 3 when spectrum = true");
    assert.equal(p.lineOf(rows[5]), "steps: int = 20");
    // a typed "when x = y" is not doubled
    assert.equal(p.lineOf({ name: "z", kind: "int", when: "when spectrum = true" }),
                 "z: int when spectrum = true");
    assert.ok(p.schemaOf(rows).endsWith("steps: int = 20\n"));
    assert.equal(p.schemaOf(rows).split("\n").length, FIELDS.length + 1);
});

test("the preset chooser keeps its name on a frontend that renames duplicates", async () => {
    const { p } = await harness();
    const node = new Node({});
    // as declared by the server: a plain text widget, not yet swapped
    Object.assign(field(node, "preset"), { type: "text", value: "vanilla" });
    const combo = p.asDropdown(node, "preset", () => ["custom", "vanilla"]);
    assert.equal(combo.name, "preset", "issue #12: renamed to preset#1 when the old widget was still there");
    assert.equal(combo.type, "combo");
    assert.equal(combo.value, "vanilla", "the value carried over");
    assert.equal(node.widgets.filter((w) => w.name === "preset" || w.name === "preset#1").length, 1);
    assert.equal(node.widgets.indexOf(combo), 1, "in the old widget's place");
});

// issue #12 diagnostics: a node that cannot build says why, and is not
// left believing it is built -- nor re-asking the server every frame
test("a build that throws is recorded, not latched, and not retried per frame", async () => {
    const { p, asked } = await harness();
    const node = new Node({});
    const addWidget = node.addWidget.bind(node);
    let fail = true;
    node.addWidget = (...args) => {
        if (fail) throw new TypeError("another pack broke addWidget");
        return addWidget(...args);
    };
    await p.rebuild(node, true);
    assert.equal(node.__obvpmBuilt, null, "not left marked as built");
    assert.match(node.__obvpmBuildError, /another pack broke addWidget/);
    assert.ok(node.__obvpmTrace.some((e) => e.e === "build-threw" && /another pack/.test(e.d)));
    // the draw loop calls rebuild(node) every frame: no request for a
    // schema whose build just failed
    const before = asked.schema;
    for (let frame = 0; frame < 5; frame++) await p.rebuild(node);
    assert.equal(asked.schema, before, "no request per frame");
    // a forced retry (the spaced ones are forced) builds once it can
    fail = false;
    await p.rebuild(node, true);
    await tick();
    assert.equal(node.__obvpmBuildError, null);
    assert.ok(field(node, "steps"), "the fields are there");
    assert.ok(node.__obvpmTrace.some((e) => e.e === "built"));
});

// issue #12, the reporter's trace: on a canvas that redraws every frame
// (a pack animating), the draw loop's rebuild() during each request was
// taken as "a rebuild was asked for meanwhile", the answer discarded,
// and the node asked forever without building.
test("a canvas redrawing during the request does not discard its answer", async () => {
    const { p, asked } = await harness();
    const node = new Node({});
    const flight = p.rebuild(node);
    // a frame or two while the answer is in the air, every time
    for (let frame = 0; frame < 50 && node.__obvpmBuilding !== false; frame++) {
        void p.rebuild(node);
        await tick();
    }
    await flight;
    assert.equal(node.__obvpmBuilt, field(node, "schema").value, "built");
    assert.ok(field(node, "steps"), "the fields are there");
    assert.equal(asked.schema, 1, "asked once");
    assert.ok(!node.__obvpmTrace.some((e) => e.e === "stale"), "no answer discarded");
});

test("a different schema, or a forced build, during the request still asks again", async () => {
    // the load case: the default schema is in the air when the workflow's
    // own arrives (onConfigure) -- that answer must not be built from
    for (const how of ["new text", "forced"]) {
        const { p, asked } = await harness();
        const node = new Node({});
        const flight = p.rebuild(node);
        if (how === "new text") {
            field(node, "schema").value += "\nsteps: int = 20";
            void p.rebuild(node);
        } else {
            void p.rebuild(node, true);
        }
        await flight;
        await tick(); await tick();
        assert.equal(asked.schema, 2, how + ": asked again");
        assert.equal(node.__obvpmBuilt, field(node, "schema").value, how + ": built from the current text");
    }
});

test("a node its graph does not hold is recorded with why, and not re-asked per frame", async () => {
    const { p, asked } = await harness();
    const node = new Node({});
    const other = { type: "KSampler" };
    node.graph = { getNodeById: () => other, _nodes: [node, other] };
    other.id = node.id;
    await p.rebuild(node, true);
    const ev = node.__obvpmTrace.find((e) => e.e === "not-live");
    assert.ok(ev, "traced");
    assert.match(ev.d, /ANOTHER node under id 1 \(KSampler\).*this id in the graph: 2/);
    const before = asked.schema;
    for (let frame = 0; frame < 5; frame++) await p.rebuild(node);
    assert.equal(asked.schema, before, "no request per frame");
    assert.equal(field(node, "steps"), undefined, "no fields built on it");
});
